pkg/controllers/drift/controller.go

```diff
@@ -0,0 +1,120 @@
+
+	provisioner := &v1alpha5.Provisioner{}
+	if err := r.kubeClient.Get(ctx, req.NamespacedName, provisioner); err != nil {
+		return reconcile.Result{}, nil
+	}
+
+	nodes := v1.NodeList{}
+	if err := r.kubeClient.List(ctx, &nodes, client.MatchingLabels{v1alpha5.ProvisionerNameLabelKey: req.Name}); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	alreadyDriftedNodes := lo.Filter(nodes.Items, func(node v1.Node, _ int) bool {
+		_, drifted := node.Annotations[v1alpha5.DriftedAnnotationKey]
+		return drifted
+	})
+
+	if len(alreadyDriftedNodes) == len(nodes.Items) {
+		return reconcile.Result{}, nil
+	}
+
+	driftedNodes := r.cloudProvider.GetDriftedNodes(ctx, provisioner, nodes.Items)
+	for _, driftedNode := range driftedNodes {
+		//Check for already drifted node and continue
+		if _, ok := driftedNode.Annotations[v1alpha5.DriftedAnnotationKey]; ok {
+			continue
+		}
+
+		stored := driftedNode.DeepCopy()
+		driftedNode.Annotations[v1alpha5.DriftedAnnotationKey] = "true"
+		if err := r.kubeClient.Patch(ctx, &driftedNode, client.MergeFrom(stored)); err != nil {
+			return reconcile.Result{}, fmt.Errorf("patching node, %w", err)
+		}
+		logging.FromContext(ctx).Infof("Marked node %s as Drifted", driftedNode.Name)
+	}
+
+	//Todo: Decide time for periodic reconciliation
+	return reconcile.Result{RequeueAfter: 30 * time.Minute}, nil
+}
+
+func (c DriftController) Builder(ctx context.Context, m manager.Manager) operatorcontroller.Builder {
+	return controllerruntime.
+		NewControllerManagedBy(m).
+		Named(controllerName).
+		For(&v1alpha5.Provisioner{}).
+		Watches(
+			//Todo:RemoveComment: Not sure if this is necessary, or if primary watch should be AwsNodeTemplate,
+			//but I think its better to reconcile Provisioner as getInstanceType in cloudprovider is deeply rooted
+			//in provisioner, so we do need the reference of it, even if we primarily reconcile AwsNodeTemplate
+			&source.Kind{Type: &v1alpha1.AWSNodeTemplate{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) []reconcile.Request {
+				reconcileList := []reconcile.Request{}
+				x, ok := o.(*v1alpha1.AWSNodeTemplate)
+				if !ok {
+					logging.FromContext(ctx).Debugf("Not able to cast AWSNodeTemplate in EnqueueRequestsFromMapFunc")
+					return reconcileList
+				}
+				provisionerList := &v1alpha5.ProvisionerList{}
+				err := c.kubeClient.List(ctx, provisionerList)
+				if err != nil {
+					//log
+					logging.FromContext(ctx).Errorf("listing provisioners for AWSNodeTemplate reconciliation %w", err)
+					return reconcileList
+				}
+				for _, provisioner := range provisionerList.Items {
+					if provisioner.Spec.ProviderRef != nil && provisioner.Spec.ProviderRef.Name == x.Name {
+						reconcileList = append(reconcileList, reconcile.Request{NamespacedName: types.NamespacedName{Name: provisioner.GetName()}})
+					}
+				}
+				return reconcileList
+			}),
+		).WithOptions(controller.Options{MaxConcurrentReconciles: 10})
+}
```

Comment:
Is there any particular reason, if primary watch should be for the nodes instead of the provisioner ? 
I think it's better to provide the cloudprovider a collection of nodes to process on together rather than individual nodes. In future we might have a case where we need to query node attributes from the cloudprovider(for example DescibeInstances), and collection of nodes gives us the opportunity to batch them all in a single api call.

Reply:
To me, the source of truth for the state of the a is the `Provisioner` or `AwsNodeTemplate` and not the actual Node itself. 
Please correct me if I am wrong :)